simba/schedule.py

```diff
@@ -285,10 +301,170 @@ def assign_vehicles(self):
+            except KeyError:
+                logging.warning(f"Rotation {rot.id} ends at a non electrified station.")
+                station_is_electrified = False
+
+            # filter vehicles of the right vehicle type and right location
+            standing_vehicles = list(filter(lambda x: vt_ct in x[0] * (x[1] == rot.departure_name),
+                                            all_standing_vehicles))
+
+            def partial_function(v_id_deps: tuple):
+                """ get the partial function of soc_at_departure_time
+
+                :param v_id_deps: vehicle_id and depot name
+                :type v_id_deps: tuple
+                :returns: partial function
+                :rtype: Callable
+                """
+                return soc_at_departure_time(v_id_deps, departure_time, vehicle_data, self.stations,
+                                             charge_curves, args)
+
+            # Add the expected soc at departure time, if greedy charging would be applied
+            socs = list(map(lambda x: partial_function(x), standing_vehicles))
+            # Pair with socs
+            standing_vehicles_w_soc = [(*v, k) for v, k in zip(standing_vehicles, socs)]
+
+            standing_vehicles_w_soc = sorted(standing_vehicles_w_soc, key=lambda x: x[-1])
+            consumption_soc = rot.calculate_consumption() / self.vehicle_types[vt][ct]["capacity"]
+            for vehicle_id, depot, soc in standing_vehicles_w_soc:
+                # end soc of vehicle if it services this rotation
+                end_soc = soc - consumption_soc
+
+                if end_soc > 0 or soc >= initial_soc or not station_is_electrified:
+                    # assign vehicles if end_soc is positive or
+                    # if soc has initial_soc.
+                    # in this case generating a new vehicle would not make a difference.
+                    # or station is not electrified. in this case vehicles should not strand there
+                    # even if rotations are not possible
+                    rot.vehicle_id = vehicle_id
+                    all_standing_vehicles.remove((vehicle_id, depot))
+                    vehicle_data[vehicle_id] = {"soc": end_soc,
+                                                "arrival_time": rot.arrival_time}
+                    break
+            else:
+                # no vehicle available for dispatch, generate new one
+                vehicle_type_counts[vt_ct] += 1
+                vehicle_id = f"{vt_ct}_{vehicle_type_counts[vt_ct]}"
+                rot.vehicle_id = vehicle_id
+                end_soc = initial_soc - consumption_soc
+                vehicle_data[vehicle_id] = {"soc": end_soc,
+                                            "arrival_time": rot.arrival_time}
+
+            rotations_in_progress.append(rot)
+            try:
+                next_rot = rotations[k + 1]
+                departure_time = next_rot.departure_time
+            except IndexError:
+                pass
```

Comment:
not sure what is happening here, needs more comments. Is it just checking that a next rotation exists or is it important that departure_time is set (not really used elsewhere)?

Reply:
it checks if a next rotation exists to get the optional value of departure_time. I changed the except clause to a break instead of a pass. Both do more or less the same here. This should make it more explicit and alows for removing the else: clause